Dice app for java (#2625)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/java/getting-started.md

---
title: Getting Started
description: Get telemetry from a client-server app in less than 5 minutes!
spelling: cSpell:ignore helloworld javaagent springframework autoreconfigure
spelling: cSpell:ignore rolldice aarch Nanos darwin autoconfigure webmvc
weight: 1
---

This page will show you how to get started with OpenTelemetry in Java.

You will learn how you can instrument a simple java application automatically,
in such a way that traces, metrics and logs are emitted to the console.

## Prerequisites

Ensure that you have the following installed locally:

- Java JDK
- [Gradle](https://gradle.org/)

## Example Application

The following example uses a basic spring boot application.

### Dependencies

Create a folder called `java-simple` and within that folder, create a file

> called `build.gradle` with the following content:

Reviewer: ```suggestion
called `build.gradle.kts` with the following content:
```

> ```gradle
> plugins {
>   id 'java'
>   id 'org.springframework.boot' version '3.0.5'
>   id 'io.spring.dependency-management' version '1.1.0'
> }
> 
> 
> sourceSets {
>         main.java.srcDirs = ['.']
> }
> 
> repositories {
>   mavenCentral()
> }
> 
> dependencies {
>   implementation 'org.springframework.boot:spring-boot-starter-web',
>                  'org.apache.logging.log4j:log4j-core:2.17.2'
> }
> ```

Reviewer: ```suggestion
```kotlin
plugins {
  id("java")
  id("org.springframework.boot") version "3.0.6"
  id("io.spring.dependency-management") version "1.1.0"
}

sourceSets {
  main {
    java.setSrcDirs(setOf("."))
  }
}

repositories {
  mavenCentral()
}

dependencies {
  implementation("org.springframework.boot:spring-boot-starter-web")
}
```
```

### Code

In that same folder, create a file called `DiceApplication.java` and add the
following code to the file:

```java

> package otel;
> import org.springframework.boot.SpringApplication;
> import org.springframework.boot.Banner;
> import org.springframework.boot.autoconfigure.SpringBootApplication;
> @SpringBootApplication
> public class DiceApplication {
>   public static void main(String[] args) {
>     SpringApplication app = new SpringApplication(DiceApplication.class);
>     app.setBannerMode(Banner.Mode.OFF);
>     app.run(args);
>   }
> }

Reviewer: ```suggestion
package otel;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.Banner;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class DiceApplication {
  public static void main(String[] args) {
    SpringApplication app = new SpringApplication(DiceApplication.class);
    app.setBannerMode(Banner.Mode.OFF);
    app.run(args);
  }
}
```

```

Create another file called `RollController.java` and add the following code to
the file:

```java

> package otel;
> import org.springframework.web.bind.annotation.GetMapping;
> import org.springframework.web.bind.annotation.RequestParam;
> import org.springframework.web.bind.annotation.RestController;
> import org.apache.logging.log4j.LogManager;
> import org.apache.logging.log4j.Logger;
> import java.util.Optional;
> 
> 
> @RestController
> public class RollController {
>   private static final Logger logger = LogManager.getLogger(RollController.class);
>   @GetMapping("/rolldice")
>   public String index(@RequestParam("player") Optional<String> player) {
>     int result = this.getRandomNumber(-1, 6);
>     if(player.isPresent()) {
>       logger.info(player.get() + " is rolling the dice: " + result);
>     } else {
>       logger.info("Anonymous player is rolling the dice: " + result);
>     }
>     return Integer.toString(result);
>   }
>     public int getRandomNumber(int min, int max) {
>         return (int) ((Math.random() * (max - min)) + min);
>     }
> }

Reviewer: ```suggestion
package otel;

import java.util.Optional;
import java.util.concurrent.ThreadLocalRandom;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class RollController {
  private static final Logger logger = LoggerFactory.getLogger(RollController.class);

  @GetMapping("/rolldice")
  public String index(@RequestParam("player") Optional<String> player) {
    int result = this.getRandomNumber(1, 6);
    if (player.isPresent()) {
      logger.info("{} is rolling the dice: {}", player.get(), result);
    } else {
      logger.info("Anonymous player is rolling the dice: {}", result);
    }
    return Integer.toString(result);
  }

  public int getRandomNumber(int min, int max) {
    return ThreadLocalRandom.current().nextInt(min, max + 1);
  }
}
```

```

To test the application, build and run it:

```console
gradle assemble

> java -jar ./java-simple.jar

Reviewer: ```suggestion
java -jar ./build/libs/java-simple.jar
```

```

## Instrumentation

Next, you'll use a [Java agent to automatically instrument](../automatic) the
application at launch time. While you can [configure the Java agent][] in a
number of ways, the steps below use environment variables.

1. Download [opentelemetry-javaagent.jar][] from [Releases][] of the
   `opentelemetry-java-instrumentation` repo. The JAR file contains the agent
   and all automatic instrumentation packages:

   ```console
   curl -L -O https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/latest/download/opentelemetry-javaagent.jar
   ```

   {{% alert color="info" %}}<i class="fas fa-edit"></i> Take note of the path
   to the JAR file.{{% /alert %}}

2. Set and export variables that specify the Java agent JAR and a [console
   exporter][], using a notation suitable for your shell/terminal environment
   &mdash; we illustrate a notation for bash-like shells:

   ```console

>    $ export JAVA_OPTS="-javaagent:PATH/TO/opentelemetry-javaagent.jar" \

Reviewer: AFAIK `JAVA_TOOL_OPTIONS` is the universally recognized env var name -- at least that what's been working for me (temurin 17)
```suggestion
   $ export JAVA_TOOL_OPTIONS="-javaagent:PATH/TO/opentelemetry-javaagent.jar" \
```

Author: Good call, I ran into that issue myself but forgot to fix that!

     OTEL_TRACES_EXPORTER=logging \
     OTEL_METRICS_EXPORTER=logging \
     OTEL_LOGS_EXPORTER=logging
   ```

   {{% alert title="Important" color="warning" %}}Replace `PATH/TO` above, with
   your path to the JAR.{{% /alert %}}

3. Run your **application** once again:

   ```console

>    $ java -jar ./java-simple.jar

Reviewer: ```suggestion
   $ java -jar ./build/libs/java-simple.jar
```

- Learn about [manual instrumentation][] and try out more
  [examples]({{% relref examples %}}).

> ## Troubleshooting

Reviewer: There already is a troubleshooting section here: https://opentelemetry.io/docs/instrumentation/java/automatic/#troubleshooting

Author: I put it here to be consistent with the JavaScript documentation, but let me remove it for now, we don't need to have it twice
